proof-systems/src/darlin/t_dlog_acc_marlin/mod.rs

```diff
@@ -0,0 +1,734 @@
+            max_degree, num_constraints, num_variables,
+        )
+        });
+
+        let srs = PC::<G1, D>::setup(max_degree).map_err(Error::from_pc_err);
+        end_timer!(setup_time);
+        srs
+    }
+
+    /// The circuit-specific setup. Given a circuit `c` and a committer_key of the polynomial
+    /// commitment scheme, generate the key material for the circuit. The latter is split into
+    /// a prover key and a verifier key.
+    pub fn circuit_specific_setup<C: ConstraintSynthesizer<G1::ScalarField>>(
+        committer_key: &<PC<G1, D> as PolynomialCommitment<G1>>::CommitterKey,
+        c: C,
+    ) -> Result<
+        (ProverKey<G1, G2, D>, VerifierKey<G1, G2, D>),
+        Error<<PC<G1, D> as PolynomialCommitment<G1>>::Error>,
+    > {
+        let index_time = start_timer!(|| "Marlin::Index");
+
+        let index = IOP::<G1, G2>::index(c)?;
+
+        end_timer!(index_time);
+
+        let commit_time = start_timer!(|| "Commit to index polynomials");
+
+        let (index_comms, index_comm_rands): (_, _) =
+            <PC<G1, D> as PolynomialCommitment<G1>>::commit_vec(committer_key, index.iter(), None)
+                .map_err(Error::from_pc_err)?;
+        end_timer!(commit_time);
+
+        let index_comms = index_comms
+            .into_iter()
+            .map(|c| c.commitment().clone())
+            .collect();
+
+        // Compute the commitments of the Lagrange polynomials over the input domain.
+        // They are included into the verifier key in order to help the verifier check that the
+        // polynomial behind the commitment of the input poly provided by the prover is indeed the
+        // input poly.
+        let domain_x = get_best_evaluation_domain(index.index_info.num_inputs).unwrap();
+        let lagrange_comms: Vec<_> = domain_x
+            .elements()
+            .into_iter()
+            .map(|y| domain_x.slice_lagrange_kernel(y))
+            .map(|poly| {
+                <PC<G1, D> as PolynomialCommitment<G1>>::commit(committer_key, &poly, false, None)
+                    .unwrap()
+                    .0
+            })
+            .collect();
+
+        let index_vk = VerifierKey {
+            index,
+            index_comms,
+            lagrange_comms,
+        };
+
+        let index_pk = ProverKey {
+            index_vk: index_vk.clone(),
+            index_comm_rands,
+        };
+
+        Ok((index_pk, index_vk))
+    }
+
+    fn fiat_shamir_rng_init(
+        pc_vk: &<PC<G1, D> as PolynomialCommitment<G1>>::VerifierKey,
+        index_vk: &VerifierKey<G1, G2, D>,
+        x_poly_comm: &<PC<G1, D> as PolynomialCommitment<G1>>::Commitment,
+        inner_sumcheck_acc: &DualSumcheckItem<G2, G1>,
+        dlog_acc: &DualDLogItem<G2, G1>,
+    ) -> Result<<PC<G1, D> as PolynomialCommitment<G1>>::RandomOracle, poly_commit::error::Error>
+    {
+        let fs_rng_init_seed = {
```

Comment:
We need to make this circuit friendly. The idea is to hardcode Fiat Shamir initial state, thus in the primitive we need to initialize the seed only with data that is constant regardless of the circuit.
In this case (similarly as we did in https://github.com/HorizenOfficial/ginger-lib/blob/sponge/proof-systems/src/darlin/marlin/src/lib.rs#L166 ) we can put into the seed only PROTOCOL_NAME and pc_pk.get_hash(). The rest is variable data, thus it must be explicitly immediately absorbed after having initialized the seed.
Same reasoning also for the verifier.

Reply:
Done